published/external/xdpapi.h

```diff
@@ -474,6 +486,61 @@ XdpOpenApi(
     return _Status;
 }
 
+#define _POLL_INTERVAL_MS 10
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+NTSTATUS
+XdpOpenApi(
+    _In_ UINT32 _XdpApiVersion,
+    _In_opt_ VOID *_ClientContext,
+    _In_opt_ XDP_API_ATTACH_FN *_ClientAttach,
+    _In_opt_ XDP_API_DETACH_FN *_ClientDetach,
+    _In_ const XDP_API_CLIENT_DISPATCH *_XdpApiClientDispatch,
+    _In_ const INT64 _TimeoutMs,
+    _Out_ XDP_API_CLIENT *_ApiContext,
+    _Out_ const XDP_API_PROVIDER_DISPATCH **_XdpApiProviderDispatch,
+    _Out_ const XDP_API_PROVIDER_BINDING_CONTEXT **_XdpApiProviderContext
+    )
+{
+    NTSTATUS Status;
+    INT64 TimeoutMs = MAXLONGLONG; // forever
+    if (_TimeoutMs > 0) {
+        TimeoutMs = _TimeoutMs;
+    }
+
+    KeInitializeEvent(&_ApiContext->AttachEvent, NotificationEvent, FALSE);
+
+    Status = XdpRegister(_XdpApiVersion, _ClientContext, _ClientAttach, _ClientDetach, _XdpApiClientDispatch, _ApiContext);
+    if (!NT_SUCCESS(Status)) {
+        return Status;
+    }
+
+    do {
+        LARGE_INTEGER Timeout100Ns;
+
+        Status =
+            XdpGetProviderContexts(
+                _ApiContext,
+                _XdpApiProviderDispatch,
+                _XdpApiProviderContext);
+        if (NT_SUCCESS(Status)) {
+            break;
+        }
+
+        Timeout100Ns.QuadPart = -1 * Int32x32To64(_POLL_INTERVAL_MS, 10000);
```

Comment:
is there a need to poll with the event? can't this code just wait for the event, with an optional timeout?

Reply:
I believe this is required because `XdpGetProviderContexts` potentially immediately returns `STATUS_DEVICE_NOT_READY`.
Or the `KeWaitForSingleObject to be placed just after `XdpRegister`?

Reply:
Yeah, I think you should be able to place with wait after XdpRegister and before XdpGetProviderContexts, so there should be no need for a loop or polling.